cabal-install-solver/src/Distribution/Solver/Modular/Message.hs

```diff
@@ -236,61 +235,37 @@ blurbOption :: ProgressAction -> QPN -> POption -> String
-data IsOrVs = Is [POption] | Vs [Ver] deriving Show
-
--- | Try to convert a list of options to a list of versions, or a list of
--- instances if any of the options is linked or installed.  Singleton lists or
--- empty lists are always converted to Is.
--- >>> tryVs [v0, v1]
--- Vs [mkVersion [0],mkVersion [1]]
--- >>> tryVs [v0]
--- Is [POption (I (mkVersion [0]) InRepo) Nothing]
--- >>> tryVs [i0, i1]
--- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-bar-0-inplace"))) Nothing,POption (I (mkVersion [1]) (Inst (UnitId "foo-bar-1-inplace"))) Nothing]
--- >>> tryVs [i0, v1]
--- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-bar-0-inplace"))) Nothing,POption (I (mkVersion [1]) InRepo) Nothing]
--- >>> tryVs [v0, i1]
--- Is [POption (I (mkVersion [0]) InRepo) Nothing,POption (I (mkVersion [1]) (Inst (UnitId "foo-bar-1-inplace"))) Nothing]
--- >>> tryVs [i0]
--- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-bar-0-inplace"))) Nothing]
--- >>> tryVs []
--- Is []
-tryVs :: [POption] -> IsOrVs
-tryVs xs@[] = Is xs
-tryVs xs@[_] = Is xs
-tryVs xs
-    | any (\(POption (instI -> b0) (isJust -> b1)) -> b0 || b1) xs = Is xs
-    | otherwise =
-      let (vs, is) = L.partition ((== InRepo) . snd) [(v, l) | POption i _ <- xs, let I v l = i]
-      in if null is then Vs (fst `map` vs) else Is xs
-
--- | Shows a list of versions in a human-friendly way, abbreviated. Shows a list
--- of instances in full.
--- >>> showIsOrVs foobarQPN $ tryVs [v0, v1]
+-- | Shows a mixed list of instances and versions in a human-friendly way,
+-- abbreviated.
+-- >>> showIsOrVs foobarQPN [v0, v1]
 -- "foo-bar; 0, 1"
--- >>> showIsOrVs foobarQPN $ tryVs [v0]
+-- >>> showIsOrVs foobarQPN [v0]
 -- "foo-bar-0"
--- >>> showIsOrVs foobarQPN $ tryVs [i0, i1]
--- "foo-bar-0/installed-inplace, foo-bar-1/installed-inplace"
--- >>> showIsOrVs foobarQPN $ tryVs [i0, v1]
--- "foo-bar-0/installed-inplace, foo-bar-1"
--- >>> showIsOrVs foobarQPN $ tryVs [v0, i1]
--- "foo-bar-0, foo-bar-1/installed-inplace"
--- >>> showIsOrVs foobarQPN $ tryVs []
+-- >>> showIsOrVs foobarQPN [i0, i1]
+-- "foo-bar; 0/installed-inplace, 1/installed-inplace"
+-- >>> showIsOrVs foobarQPN [i0, v1]
+-- "foo-bar; 0/installed-inplace, 1"
+-- >>> showIsOrVs foobarQPN [v0, i1]
+-- "foo-bar; 0, 1/installed-inplace"
+-- >>> showIsOrVs foobarQPN []
 -- "unexpected empty list of versions"
-showIsOrVs :: QPN -> IsOrVs -> String
-showIsOrVs _ (Is []) = "unexpected empty list of versions"
-showIsOrVs q (Is xs) = L.intercalate ", " (showOption q `map` xs)
-showIsOrVs q (Vs xs) = showQPN q ++ "; " ++ L.intercalate ", " (showVer `map` xs)
+showIsOrVs :: QPN -> [POption] -> String
+showIsOrVs _ [] = "unexpected empty list of versions"
+showIsOrVs q [x] = showOption q x
+showIsOrVs q xs = showQPN q ++ "; " ++ (L.intercalate ", "
+  [if isJust linkedTo
+    then showOption q x
+    else showI i -- Don't show the package, just the version
+  | x@(POption i linkedTo) <- xs
+  ])
```

Comment:
Could you please add a test for the linked case?

Reply:
Yes, I've added these two;

```haskell
-- >>> showOptions foobarQPN [k2, k3]
-- "foo-bar; foo-bar~>foo-bar-2, foo-bar~>foo-bar-3"
-- >>> showOptions foobarQPN [v0, i1, k3]
-- "foo-bar; 0, 1/installed-inplace, foo-bar~>foo-bar-3"
```

We can't abbreviate the linked ones can we?

I haven't dropped the prefix when all are linked as in the first example above.

Reply:
Yes, I can't think of a clear way to abbreviate the linked ones, especially since the two `PackagePath`s are always different when the solver links packages, so there is never a common prefix. I also think that it would be better to use different `PackagePath`s for the `Maybe PackagePath` field of the `POption` for `k1` and `k2`, to make it more realistic.

Reply:
Yes that is how it turned out;

```
-- >>> showOptions foobarQPN [k1, k2]
-- "foo-bar; foo-bar~>bazqux.foo-bar-1, foo-bar~>bazqux.foo-bar-2"
-- >>> showOptions foobarQPN [v0, i1, k2]
-- "foo-bar; 0, 1/installed-inplace, foo-bar~>bazqux.foo-bar-2"
```